### PR TITLE
[Hub Generated] Review request for Microsoft.Insights to add version stable/2022-06-15

### PR DESCRIPTION
**Pull request by avjerafi**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/avjerafi/azure-rest-api-specs/dev-monitor-Microsoft.Insights-2022-06-15-sqr...main/).
## ARM API Information (Control Plane)

Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by avjerafi** [comment]
> Hi @avjerafi, Your PR has some issues. Please fix the CI sequentially by following the order of `Avocado, semantic validation, model validation, breaking change, lintDiff`. If you have any questions, please post your questions in this channel https://aka.ms/swaggersupport.
> 
> Task	How to fix	Priority
> Avocado	[Fix-Avocado](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/ci-fix.md#avocado)	High
> Semantic validation	[Fix-SemanticValidation-Error](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/ci-fix.md#semantic-validation)	High
> Model validation	[Fix-ModelValidation-Error](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/ci-fix.md#model-validation)	High
> LintDiff	[Fix-LintDiff](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/ci-fix.md#linter-validation)	high
> If you need further help, please feedback via [swagger feedback](https://aka.ms/swaggerfeedback).

@changlong-liu Please advise regarding the apparent validation errors. I believe they are unjustified:

- I've executed all these validations locally for our resource type ("ScheuledQueryRules") prior to submitting this PR.
- Following these validation errors I've re-executed them locally again for the entire service ("monitor") and reviewed the results:
  - Avocado + Semantic + Model validations: Local executions return 0 errors which are actually related to the submitted changes (the ScheduledQueryRules resource type).
  - Prettier: Executed for the entire monitor service and got 0 changes.

**Comment by avjerafi** [comment]
@changlong-liu  Please note: this new api version introduces no changes to the spec itself.

Beginning with this api version, our service will require that all Action Groups which are references in Log Search Alert rules must already be provisioned whenever a rule is added/updated - a restriction which was not imposed until now.

This is necessary in order to comply with the Schrems II requirement of looking up and tracking the location of all referenced Action Groups.

This new constraint has the potential to break existing deployments for some customers. The introduction of this new api version is intended to let customers opt-in to this.

File: specification/monitor/resource-manager/readme.md
Comment (changlong-liu) [inline specification/monitor/resource-manager/readme.md:38]: Hi @avjerafi , is the 'package-2022-07' correct tag? I don't see it's definition in this readme file

**Comment by changlong-liu** [comment]
> @changlong-liu Please note: this new api version introduces no changes to the spec itself.
> 
> Beginning with this api version, our service will require that all Action Groups which are references in Log Search Alert rules must already be provisioned whenever a rule is added/updated - a restriction which was not imposed until now.
> 
> This is necessary in order to comply with the Schrems II requirement of looking up and tracking the location of all referenced Action Groups.
> 
> This new constraint has the potential to break existing deployments for some customers. The introduction of this new api version is intended to let customers opt-in to this.

Hi @avjerafi , the new api version need arm review first, and then we will go to SDK breaking change review.

File: specification/monitor/resource-manager/readme.md
Comment (avjerafi) [reply specification/monitor/resource-manager/readme.md]: @changlong-liu That was not correct. Pushed a fix.

**Comment by avjerafi** [comment]
Hi @changlong-liu,
I'm still waiting for ARM review. Do you know can I ping in order to expedite this? We would like to have this api version approved within this week if possible.

**Comment by changlong-liu** [comment]
Hi @avjerafi , would you also taking a look at the avocado issues: https://github.com/Azure/azure-rest-api-specs/pull/19652/checks?check_run_id=7210901264.

For arm review, you may ping Azure Resource Manager RP API Review <armapireview@microsoft.com> if in urgent.

File: specification/monitor/resource-manager/readme.md
Comment (changlong-liu) [inline specification/monitor/resource-manager/readme.md:38]: Hi @avjerafi , I don't find "package-2022-06-15-only" defined in readme.md, this should be the reason that make SDK generations failed.

**Comment by changlong-liu** [comment]
Hi @ruowan , the LintDiff error seems come from the common-types v2 which is referenced by this PR: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v2/types.json#L17 .

Would you help to confirm whether it blocks this PR?

File: specification/monitor/resource-manager/readme.md
Comment (avjerafi) [reply specification/monitor/resource-manager/readme.md]: @changlong-liu You're right. Thanks! Pushed a fix.

**Comment by avjerafi** [comment]
Hi @changlong-liu,
Can you please add clarity to this new MISSING_APIS_IN_DEFAULT_TAG validation error in avocado?
Do I understand correctly that the api version which is missing is 2018-04-16? This is the api version I'm replacing in the default tag in this PR. Shouldn't it be missing?
Please advise. Thanks!

**Comment by changlong-liu** [comment]
> Hi @changlong-liu, Can you please add clarity to this new MISSING_APIS_IN_DEFAULT_TAG validation error in avocado? Do I understand correctly that the api version which is missing is 2018-04-16? This is the api version I'm replacing in the default tag in this PR. Shouldn't it be missing? Please advise. Thanks!

Hi @avjerafi , the full description can be found at https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1704911&view=logs&j=22cb5f76-5058-5a9d-2377-d04e5451a5f2&t=75f8d99b-764d-5b43-90d5-e9f774389f08.

It's complaining about the path '/subscriptions/{}/resourcegroups/{}/providers/Microsoft.Insights/scheduledQueryRules' once existed in 2018-04-16 no longer exist in the latest swagger.
Would you confirm whether it's expected or not?

**Comment by avjerafi** [comment]
Hi @changlong-liu,
Thanks for the clarification.
I verified that all 3 paths in the specs of both api versions 2018-04-16 and the new 2022-06-15 remained exactly the same:
/subscriptions/{subscriptionId}/providers/Microsoft.Insights/scheduledQueryRules
/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Insights/scheduledQueryRules
/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Insights/scheduledQueryRules/{ruleName}

**Comment by changlong-liu** [comment]
> Hi @changlong-liu, Thanks for the clarification. I verified that all 3 paths in the specs of both api versions 2018-04-16 and the new 2022-06-15 remained exactly the same: /subscriptions/{subscriptionId}/providers/Microsoft.Insights/scheduledQueryRules /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Insights/scheduledQueryRules /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Insights/scheduledQueryRules/{ruleName}

Hi @ruowan , would you help to take a look whether this is a false alarm?

**Comment by ruowan** [comment]
Thanks for feedback, it's a false alarm. We will fix avocado issue. For this PR, I added approved-Avocado, it won't block your PR process.

**Comment by avjerafi** [comment]
Thank you @ruowan for your swift response!
Regarding the LintDiff validation errors:
As far as I can tell, these were not triggered by the changes in this PR. Is it possible to unblock this api version from being released?

**Comment by changlong-liu** [comment]
Hi @avjerafi , for the "Swagger LintDiff" and "SDK Track2 Validation" Error, both of them are complaining about incompatible Schema. 
They are about defining/referencing multiple models which has same name, and they are incompatible,
Would you please check?

**Comment by avjerafi** [comment]
> Hi @avjerafi , for the "Swagger LintDiff" and "SDK Track2 Validation" Error, both of them are complaining about incompatible Schema. They are about defining/referencing multiple models which has same name, and they are incompatible, Would you please check?

Hi @changlong-liu,
Yes, I can see the cause: there are duplicate type definitions which are not identical. But these involve several resource types, which are not handled by my team and I cannot modify. The error messages include a suggestion to use the "modelerfour.lenient-model-deduplication" setting to temporarily suppress these comparisons. How do I set this? I could not find this setting being used anywhere the swagger spec.

File: specification/monitor/resource-manager/readme.md
Comment (changlong-liu) [inline specification/monitor/resource-manager/readme.md:39]: > > Hi @avjerafi , for the "Swagger LintDiff" and "SDK Track2 Validation" Error, both of them are complaining about incompatible Schema. They are about defining/referencing multiple models which has same name, and they are incompatible, Would you please check?
> 
> Hi @changlong-liu, Yes, I can see the cause: there are duplicate type definitions which are not identical. But these involve several resource types, which are not handled by my team and I cannot modify. The error messages include a suggestion to use the "modelerfour.lenient-model-deduplication" setting to temporarily suppress these comparisons. How do I set this? I could not find this setting being used anywhere in the swagger spec.

You may try to adding lenient here as:
~~~
modelerfour:
  lenient-model-deduplication: true
~~~

Comment (avjerafi) [reply specification/monitor/resource-manager/readme.md:39]: Thank you @changlong-liu.

I've applied the flag and the "SDK Track2 Validation" as you directed, and the error was suppressed.

However, LintDiff validation error is still not suppressed, even though they both refer to the same issue, as far as I can tell. I cannot completely resolve this within the scope of this PR because it involves cooperation with many teams in order to align all the types used in all the different resource types.

This effort will certainly be part of the next SDK release, which we are working on these days in parallel to publishing this new api version.

Please advise.
Thanks!

File: specification/monitor/resource-manager/Microsoft.Insights/stable/2022-06-15/scheduledQueryRule_API.json
Comment (changlong-liu) [inline specification/monitor/resource-manager/Microsoft.Insights/stable/2022-06-15/scheduledQueryRule_API.json:441]: Hi @avjerafi , 

I would suggest fix the incompatible issues in this PR.
I can find that problem comes from Line 441  and Line 656 where refer to common-types.  Would you try try to not  ref  to common-types here?
I see other specs in this tag defines some Resource, maybe you can reuse these definitions?

> Thank you @changlong-liu.
> 
> I've applied the flag as you directed and the "SDK Track2 Validation" error was suppressed.
> 
> However, LintDiff validation error is still not suppressed, even though they both refer to the same issue, as far as I can tell. I cannot completely resolve this within the scope of this PR because it involves cooperation with many teams in order to align all the types used in all the different resource types.
> 
> This effort will certainly be part of the next SDK release, which we are working on these days in parallel to publishing this new api version.
> 
> Please advise. Thanks!

Comment (avjerafi) [reply specification/monitor/resource-manager/Microsoft.Insights/stable/2022-06-15/scheduledQueryRule_API.json]: Hi @changlong-liu,

- For "Resource": I've removed the "allOf" reference and inlined all the fields inside the local definition "ScheduledQueryRuleResource".
- For "ErrorResponse": I've duplicated the definition from common-types as a new local definition, renamed as the unique name "ErrorResponseDetails" in order to avoid any conflicts.

Despite that, I still get this LintDiff validation error: "model name with non-identical definitions" but I can't find which definition this error relates to.

All the local validations done on my machine passed.

Please advise.
Thanks!
Please advise.
Thanks you!

Comment (changlong-liu) [reply specification/monitor/resource-manager/Microsoft.Insights/stable/2022-06-15/scheduledQueryRule_API.json]: hI @avjerafi ,
Detail errors can be found at https://dev.azure.com/azure-sdk/590cfd2a-581c-4dcb-a12e-6568ce786175/_apis/build/builds/1708473/logs/323.
Now it's for duplicated definition of 'ErrorAdditionalInfo'.

And for the LintDiff error, you should can get reproduced in local with below command:
~~~
autorest --v2 --validation --azure-validator --semantic-validator=false --model-validator=false --message-format=json --openapi-type=arm --use=@microsoft.azure/classic-openapi-validator@1.2.4 --use=@microsoft.azure/openapi-validator@1.13.0 --tag=package-2022-06 [PATH-TO-SWAGGER-REPO]/specification/monitor/resource-manager/readme.md
~~~

Comment (avjerafi) [reply specification/monitor/resource-manager/Microsoft.Insights/stable/2022-06-15/scheduledQueryRule_API.json]: Hi @changlong-liu,
Thank you for your guidance!
I've fixed the "ErrorAdditionalInfo" reference by renaming it as well.
LintDiff validation now finally passes.
Is the PR now approvable?
Thanks!

**Comment by changlong-liu** [comment]
hi @qiaozha , @ArcturusZhang , would you help review the breaking changes in JS/GoLang? Thanks!

File: specification/monitor/resource-manager/readme.md
Comment (ArcturusZhang) [reply specification/monitor/resource-manager/readme.md:39]: We should not add this flag here, this will silently take effect in every SDK, and SDKs will not want this flag.
For Go and DotNet SDK, if your swagger must have this flag to be generated correctly, you will never GA.
Therefore if the purpose of adding this flag is just to "suppress" the track 2 validation failure, please do not do that, it only suppress the failure instead of fixing it. The failure in that pipeline is something the SDK owners would like to know and deal with. Please remove this flag and let it fail

**Review by ArcturusZhang — CHANGES_REQUESTED** [review]

**Review by ArcturusZhang — APPROVED** [review]

**Comment by changlong-liu** [comment]
Hi @avjerafi , I see "DoNotMerge" label is added. Ping me please if you want to merge this, thanks!

**Comment by avjerafi** [comment]
@changlong-liu The ARM manifest was fully deployed and I've removed the DoNotMerge flag. Please merge this PR. Thanks!

**Comment by changlong-liu** [comment]
> @changlong-liu The ARM manifest was fully deployed and I've removed the DoNotMerge flag. Please merge this PR. Thanks!

Adding @ArcturusZhang as the new assignee.

**Comment by avjerafi** [comment]
@ArcturusZhang Requesting again to please merge this PR. Thank you.

**Comment by avjerafi** [comment]
@ArcturusZhang Kind reminder: please merge this PR. Thanks!

**Review by ArcturusZhang — APPROVED** [review]

**Pull request merged**
